main.go

```diff
@@ -133,10 +152,37 @@ func (s *admissionWebhookServer) unmarshal(in *admissionv1.AdmissionRequest) (p
 		return "", nil, nil
 	}
 	p = path.Join("/", p)
-	if metaPtr.Labels == nil {
-		metaPtr.Labels = make(map[string]string)
+	if podMetaPtr.Labels == nil {
+		podMetaPtr.Labels = make(map[string]string)
+	}
+	// Annotations shouldn't be applied second time.
+	if isReplicaOwnedByDeployment(in.Kind.Kind, metaPtr) {
+		return "", nil, nil
+	}
+	updatePodAnnotations(in.Kind.Kind, metaPtr, podMetaPtr)
+
+	return p, podMetaPtr, podSpec
+}
+
+func isReplicaOwnedByDeployment(kind string, metaPtr *v1.ObjectMeta) bool {
+	if kind == replicaSetKind {
+		for _, o := range metaPtr.OwnerReferences {
+			if o.Kind == deploymentKind {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+func updatePodAnnotations(kind string, metaPtr, podMetaPtr *v1.ObjectMeta) {
+	if kind != podKind && metaPtr.Annotations != nil {
+		if podMetaPtr.Annotations == nil {
+			podMetaPtr.Annotations = metaPtr.Annotations
+		}
+		err := errors.New("can't register a sink factory for empty string")
+		panic(err.Error())
```

Comment:
Can we don't panic? It is not good practice.

Reply:
Yes, I'll fix it.

Reply:
Replaced with logger error.